### PR TITLE
Split multiline literal into templated and nontemplated versions

### PR DESCRIPTION
**Pull request by davepagurek**
Implements https://github.com/emerald-lang/emerald/issues/21

Also renamed some methods in the preprocessor that I thought might be a bit better fitting

### CODE REVIEW

**Comment by davepagurek** [comment]
Also while I was working on literals I added tests for https://github.com/emerald-lang/emerald/issues/9. I'm going to close that issue because it turns out braces in literals actually already works! :tada:

**Comment by armcburney** [comment]
Yeah, to be honest, the preprocessor had a lot of bad names because I was originally trying to adhere to the 10 line method limit for rubocop (before we updated the .yml to accommodate for 30 lines), and created a bunch of methods that didn't have the greatest names.

File: lib/emerald.rb [hunk 1]
```diff
@@ -41,9 +41,9 @@ def process(file_name, context_file_name = nil)
     end
   end
 
-  def self.convert(input, context = {})
+  def self.convert(input, context = {}, debug: false)
```
Comment (armcburney) [inline lib/emerald.rb:44]: Is this something we pass in as a parameter, or just for internal use?

File: lib/emerald/PreProcessor.rb [hunk 2]
```diff
@@ -11,16 +11,13 @@ module Emerald
   #
   class PreProcessor
     def initialize
-      @in_literal = false
-      @current_indent = 0
-      @new_indent = 0
-      @b_count = 0
-      @output = ''
+      reset
```
Comment (armcburney) [inline lib/emerald/PreProcessor.rb:14]: This is a lot cleaner, I like it.

File: lib/emerald/PreProcessor.rb [hunk 3]
```diff
@@ -92,25 +88,31 @@ def append_opening_braces(new_indent)
...
-                 end
-    end
+    def remove_indent_whitespace(line)
+      if @in_literal
+        # Ignore indent whitespace, only count post-indent as the literal,
+        # but keep any extra whitespace that might exist for literals
+        cropped = line[@current_indent..-1] || ''
+        if @raw_literal
+          # this is a fun one https://www.ruby-forum.com/topic/143645
+          cropped = cropped.gsub("\\"){ "\\\\" }
```
Comment (AndrewMcBurney) [inline lib/emerald/PreProcessor.rb]: Yeah.... alright lol. I mean, I can't think of a better way really to do it.

File: spec/functional/templating_spec.rb [hunk 4]
```diff
@@ -30,4 +30,88 @@
       EMR
     )).to eq('<h1>Hello, Dave</h1>')
   end
+
+  it 'works in multiline literals' do
+    expect(convert(
```
Comment (armcburney) [inline spec/functional/templating_spec.rb:35]: Can you put the `convert(` on a new line? The pre-commit rubocop was complaining about this for me.

Comment (armcburney) [reply spec/functional/templating_spec.rb:35]: Like this:
```
  it 'works in multiline literals' do
    expect(
      convert(
        context: {name: 'Dave'},
        input: <<~EMR,
          h1 ->
            Hello, world,
            my name is |name|
        EMR
      )
    ).to eq(whitespace_agnostic(<<~HTML))
...
```

Comment (armcburney) [reply spec/functional/templating_spec.rb:35]: Actually, I'll make a separate PR to clean up this stuff once it's merged in. I wanted to have a 'cleanup' PR which addresses a few things in the codebase. I'll approve the changes - nice job Dave!! 💯 💯 💯

**Review by armcburney — APPROVED** [review]

File: lib/emerald.rb (hunk 1 above)
Comment (davepagurek) [reply lib/emerald.rb:44]: Just internal right now

Comment (davepagurek) [reply lib/emerald.rb:44]: We probably want to make good user-facing errors at some point though

**Pull request merged**